Hazel/Udp/UdpConnection.Fragmented.cs

```diff
@@ -0,0 +1,241 @@
+            MtuChanged?.Invoke();
+
+            if (_mtuIndex < PossibleMtu.Length - 1)
+            {
+                MtuTest((byte)(index + 1));
+            }
+        }
+
+        private void MtuTestMessageReceive(MessageReader message)
+        {
+            message.Position = message.Length - 2;
+            var mtu = message.ReadUInt16();
+
+            if (mtu != message.Length)
+            {
+                return;
+            }
+
+            ProcessReliableReceive(message.Buffer, 1, out _);
+        }
+
+        private const byte FragmentHeaderSize = sizeof(byte) + sizeof(ushort) + sizeof(ushort) + sizeof(ushort);
+
+        protected void FragmentedSend(byte sendOption, byte[] data, Action ackCallback, bool includeHeader)
+        {
+            var length = includeHeader ? data.Length + 1 : data.Length;
+
+            var id = (ushort)Interlocked.Increment(ref _lastFragmentedId);
+            var fragmentSize = Mtu;
+            var fragmentDataSize = fragmentSize - FragmentHeaderSize;
+            var fragmentsCount = (int)Math.Ceiling(length / (double)fragmentDataSize);
+
+            if (fragmentsCount >= ushort.MaxValue)
+            {
+                throw new HazelException("Too many fragments");
+            }
+
+            var acksReceived = 0;
+
+            for (ushort i = 0; i < fragmentsCount; i++)
+            {
+                var dataLength = Math.Min(fragmentDataSize, length - fragmentDataSize * i);
+                var buffer = new byte[dataLength + FragmentHeaderSize];
+
+                buffer[0] = (byte)UdpSendOption.Fragment;
+
+                AttachReliableID(buffer, 1, () =>
+                {
+                    acksReceived++;
+
+                    if (acksReceived >= fragmentsCount)
+                    {
+                        ackCallback?.Invoke();
+                    }
+                });
+
+                buffer[3] = (byte)fragmentsCount;
+                buffer[4] = (byte)(fragmentsCount >> 8);
+
+                buffer[5] = (byte)id;
+                buffer[6] = (byte)(id >> 8);
+
+                var includingHeader = i == 0 && includeHeader;
+                if (includingHeader)
+                {
+                    buffer[7] = sendOption;
+                }
+
+                Buffer.BlockCopy(data, fragmentDataSize * i - (includingHeader ? 0 : 1), buffer, FragmentHeaderSize + (includingHeader ? 1 : 0), dataLength - (includingHeader ? 1 : 0));
+
+                WriteBytesToConnection(buffer, buffer.Length);
+            }
+        }
+
+        protected void FragmentMessageReceive(MessageReader messageReader)
+        {
+            if (ProcessReliableReceive(messageReader.Buffer, 1, out var id))
+            {
+                messageReader.Position += 3;
+
+                var fragmentsCount = messageReader.ReadUInt16();
+                var fragmentedMessageId = messageReader.ReadUInt16();
+
+                lock (_fragmentedMessagesReceived)
```

Comment:
I think you should use the ConcurrentDictionary locking pattern in UdpConnectionListener where you tryget, lock, tryadd. Then once you have a unified FragmentedMessage instance you can lock that instead. Because right here you have a global lock with a decent amount of work in it and I don't think it's necessary.

Keep in mind I have not fully thought this through and I don't personally expect this to be a hot path, so it's a somewhat low pri optimization.